expression/builtin_cast_vec_test.go

```diff
@@ -152,6 +155,58 @@ func (s *testEvaluatorSuite) TestVectorizedBuiltinCastFunc(c *C) {
 	testVectorizedBuiltinFunc(c, vecBuiltinCastCases)
 }
 
+// for issue https://github.com/pingcap/tidb/issues/16825
+func (s *testEvaluatorSuite) TestVectorizedCastStringAsDecimalWithUnsignedFlagInUnion(c *C) {
+	col := &Column{RetType: types.NewFieldType(mysql.TypeString), Index: 0}
+	baseFunc, err := newBaseBuiltinFunc(mock.NewContext(), "", []Expression{col})
+	if err != nil {
+		panic(err)
+	}
+	// set `inUnion` to `true`
+	baseCast := newBaseBuiltinCastFunc(baseFunc, true)
+	baseCast.tp = types.NewFieldType(mysql.TypeNewDecimal)
+	// set the `UnsignedFlag` bit
+	baseCast.tp.Flag |= mysql.UnsignedFlag
+	cast := &builtinCastStringAsDecimalSig{baseCast}
+
+	inputs := []*chunk.Chunk{
+		genCastStringAsDecimal(false),
+		genCastStringAsDecimal(true),
+	}
+
+	for _, input := range inputs {
+		result := chunk.NewColumn(types.NewFieldType(mysql.TypeDecimal), input.NumRows())
+		c.Assert(cast.vecEvalDecimal(input, result), IsNil)
+		for i := 0; i < input.NumRows(); i++ {
+			res, isNull, err := cast.evalDecimal(input.GetRow(i))
+			c.Assert(isNull, IsFalse)
+			c.Assert(err, IsNil)
+			c.Assert(result.GetDecimal(i).Compare(res), Equals, 0)
+		}
+	}
+}
+
+func genCastStringAsDecimal(isNegative bool) *chunk.Chunk {
+	var sign float64
+	if isNegative {
+		sign = -1
+	} else {
+		sign = 1
+	}
+
+	input := chunk.NewChunkWithCapacity([]*types.FieldType{types.NewFieldType(mysql.TypeString)}, 1024)
+	for i := 0; i < 1024; i++ {
+		d := new(types.MyDecimal)
+		f := sign * rand.Float64() * 100000
+		if err := d.FromFloat64(-f); err != nil {
```

Comment:
Why is `d.FromFloat64(-f)` instead of `d.FromFloat64(f)`?

Reply:
Oh, a stupid typo mistake. Thanks

Reply:
fixed